File: src/Painting/ImageProvider.ts
import invariant from 'ts-invariant'
import { Codec, Size } from '@UI'
import { SkiaTextDirection } from '@Skia'
import { Locale, StringBuffer, TargetPlatform } from '@Platform'
import { AssetBundle } from '@Services'
import { ImageErrorListener, ImageStream } from './ImageStream'
import { PaintingBinding } from './Binding'
import { ImageCacheStatus } from './ImageCache'

export type KeyAndErrorHandlerCallback<T> = { (key: T, handleError: ImageErrorListener): void }
export type AsyncKeyErrorHandler<T> = { (key: T, exception): Promise<void> }
export type DecoderCallback = {
  (
    bytes: Uint8Array, 
    cacheWidth?: number | null,
    cacheHeight?: number | null,
    allowUpscaling?: boolean
  ): Promise<Codec>
}

export type ImageConfigurationInitOptions = {
  bundle?: AssetBundle | null,
  devicePixelRatio?: number | null,
  locale?: Locale | null,
  textDirection?: SkiaTextDirection | null,
  size?: Size | null,
  platform?: TargetPlatform | null,
}

export type ImageConfigurationCopyOptions = ImageConfigurationInitOptions

export class ImageConfiguration {
  static Empty = new ImageConfiguration()
  public bundle: AssetBundle | null = null
  public devicePixelRatio: number | null = null
  public locale: Locale | null = null
  public textDirection: SkiaTextDirection | null = null
  public size: Size | null = null
  public platform: TargetPlatform | null = null
  
  constructor (options: ImageConfigurationInitOptions = {}) {
    options.bundle = options.bundle ?? null
    options.devicePixelRatio = options.devicePixelRatio ?? null
    options.locale = options.locale ?? null
    options.textDirection = options.textDirection ?? null
    options.size = options.size ?? null
    options.platform = options.platform ?? null

    this.bundle = options.bundle
    this.devicePixelRatio = options.devicePixelRatio
    this.locale = options.locale
    this.textDirection = options.textDirection
    this.size = options.size
    this.platform = options.platform
  }

  copyWith (options: ImageConfigurationInitOptions): ImageConfiguration  {
    return new ImageConfiguration({
      bundle: options.bundle ?? this.bundle,
      devicePixelRatio: options.devicePixelRatio ?? this.devicePixelRatio,
      locale: options.locale ?? this.locale,
      textDirection: options.textDirection ?? this.textDirection,
      size: options.size ?? this.size,
      platform: options.platform ?? this.platform,
    })
  }

  isEqual (other) {
    if (this === other) {
      return true
    }

    return (
      other instanceof ImageConfiguration &&
      other.bundle == this.bundle &&
      other.devicePixelRatio == this.devicePixelRatio &&
      other.locale == this.locale &&
      other.textDirection == this.textDirection &&
      other.size == this.size &&
      other.platform == this.platform
    )
  }
  
  toString () {
    const result = new StringBuffer()

    result.write('ImageConfiguration(');
    let hasArguments = false
    
    if (this.bundle !== null) {
      result.write(`bundle: ${this.bundle}`)
      hasArguments = true
    }
    
    if (devicePixelRatio !== null) {
      if (hasArguments) {
        result.write(', ')
      }
      result.write(`devicePixelRatio: ${this.devicePixelRatio!.toFixed(1)}`);
      hasArguments = true
    }

    if (this.locale !== null) {
      if (hasArguments) {
        result.write(', ')
      }
      result.write(`locale: ${this.locale}`)
      hasArguments = true
    }
    if (this.textDirection !== null) {
      if (hasArguments) {
        result.write(', ')
      }
      result.write(`textDirection: ${this.textDirection}`)
      hasArguments = true
    }
    if (this.size !== null) {
      if (hasArguments) {
        result.write(', ')
      }
      result.write(`size: ${this.size}`)
      hasArguments = true
    }
    if (this.platform !== null) {
      if (hasArguments) {
        result.write(', ')
      }
      // @TODO
      result.write(`platform: ${this.platform!}`)
      hasArguments = true
    }

    result.write(')')

    return result.toString()
  }
}

export abstract class ImageProvider<T> {
  abstract obtainedKey (configuration: ImageConfiguration): Promise<T>
  abstract load (key: T, decode: DecoderCallback)

  resolve (configuration: ImageConfiguration) {
    invariant(configuration !== null)

    const stream = this.createStream(configuration)

    this.createErrorHandlerAndKey(configuration, (key: T, onError: ImageErrorListener) => {
      this.resolveStreamForKey(configuration, stream, key, onError)
    }, async (key: T | null, exception) => {
         
    })

    return stream
  }

  createStream (configuration: ImageConfiguration): ImageStream {
    return new ImageStream()
  }

  obtainCacheStatus (
    configuration: ImageConfiguration,
    onError: ImageErrorListener,
  ): Promise<ImageCacheStatus | null>  {
    invariant(configuration !== null)
    
    return new Promise((resolve, reject) => {
      this.createErrorHandlerAndKey(
        configuration,
        (key: T) => {
          resolve(PaintingBinding.instance!.imageCache!.statusForKey(key))
        },
        async (key: T | null, exception) => {
          if (onError !== null) {
            onError(exception)
          } else {
            reject(exception)
          }
        },
      )
    })
    
  }

  createErrorHandlerAndKey (
    configuration: ImageConfiguration ,
    success: KeyAndErrorHandlerCallback<T>,
    error: AsyncKeyErrorHandler<T | null> ,
  ) {
    let obtainedKey: T | null
    let didError = false

    const onError = async exception => {
      if (didError) {
        return
      }

      if (!didError) {
        error(obtainedKey, exception)
      }
      didError = true
    }

    try {
      const key = this.obtainedKey(configuration)

      key.then<void>((key: T) => {
        obtainedKey = key
        try {
          success(key, onError)
        } catch (error) {
          onError(error)
        }
      }).catch(onError)
    } catch (error) {
      onError(error)
    }
  }

  resolveStreamForKey (
    configuration: ImageConfiguration, 
    stream: ImageStream, 
    key: T, 
    onError: ImageErrorListener
  ): void  {

    if (stream.completer !== null) {
      const completer = PaintingBinding.instance!.imageCache!.putIfAbsent(key, () => stream.completer!, onError,)
      invariant(completer === stream.completer)
      return
    }

    const completer = PaintingBinding.instance!.imageCache!.putIfAbsent<T>(key, () => {
      return this.load(key, PaintingBinding.instance!.instantiateImageCodec)
    }, onError)
    
    if (completer !== null) {
      stream.setCompleter(completer)
    }
  }
}


export class AssetBundleImageKey {
  public bundle: AssetBundle
  public name: string
  public scale: number

  constructor(
    bundle: AssetBundle,
    name: string,
    scale: number,
  ) {
    invariant(bundle !== null)
    invariant(name !== null)
    invariant(scale !== null)

    this.bundle = bundle
    this.name = name
    this.scale = scale
  }

  isEqual (other: AssetBundleImageKey) {
    if (other === this) {
      return true
    }
    return (
      other instanceof AssetBundleImageKey &&
      other.bundle === this.bundle &&
      other.name === this.name &&
      other.scale === this.scale
    )
  }

  toString () {
    return ''
  }
}

File: dart/classes/ImageProvider.dart
abstract class ImageProvider<T extends Object> {
  /// Abstract const constructor. This constructor enables subclasses to provide
  /// const constructors so that they can be used in const expressions.
  const ImageProvider();

  /// Resolves this image provider using the given `configuration`, returning
  /// an [ImageStream].
  ///
  /// This is the public entry-point of the [ImageProvider] class hierarchy.
  ///
  /// Subclasses should implement [obtainKey] and [load], which are used by this
  /// method. If they need to change the implementation of [ImageStream] used,
  /// they should override [createStream]. If they need to manage the actual
  /// resolution of the image, they should override [resolveStreamForKey].
  ///
  /// See the Lifecycle documentation on [ImageProvider] for more information.
  @nonVirtual
  ImageStream resolve(ImageConfiguration configuration) {
    assert(configuration != null);
    final ImageStream stream = createStream(configuration);
    // Load the key (potentially asynchronously), set up an error handling zone,
    // and call resolveStreamForKey.
    _createErrorHandlerAndKey(
      configuration,
      (T key, ImageErrorListener errorHandler) {
        resolveStreamForKey(configuration, stream, key, errorHandler);
      },
      (T? key, Object exception, StackTrace? stack) async {
        await null; // wait an event turn in case a listener has been added to the image stream.
        InformationCollector? collector;
        assert(() {
          collector = () => <DiagnosticsNode>[
                DiagnosticsProperty<ImageProvider>('Image provider', this),
                DiagnosticsProperty<ImageConfiguration>(
                    'Image configuration', configuration),
                DiagnosticsProperty<T>('Image key', key, defaultValue: null),
              ];
          return true;
        }());
        if (stream.completer == null) {
          stream.setCompleter(_ErrorImageCompleter());
        }
        stream.completer!.reportError(
          exception: exception,
          stack: stack,
          context: ErrorDescription('while resolving an image'),
          silent: true, // could be a network error or whatnot
          informationCollector: collector,
        );
      },
    );
    return stream;
  }

  /// Called by [resolve] to create the [ImageStream] it returns.
  ///
  /// Subclasses should override this instead of [resolve] if they need to
  /// return some subclass of [ImageStream]. The stream created here will be
  /// passed to [resolveStreamForKey].
  @protected
  ImageStream createStream(ImageConfiguration configuration) {
    return ImageStream();
  }

  /// Returns the cache location for the key that this [ImageProvider] creates.
  ///
  /// The location may be [ImageCacheStatus.untracked], indicating that this
  /// image provider's key is not available in the [ImageCache].
  ///
  /// The `cache` and `configuration` parameters must not be null. If the
  /// `handleError` parameter is null, errors will be reported to
  /// [FlutterError.onError], and the method will return null.
  ///
  /// A completed return value of null indicates that an error has occurred.
  Future<ImageCacheStatus?> obtainCacheStatus({
    required ImageConfiguration configuration,
    ImageErrorListener? handleError,
  }) {
    assert(configuration != null);
    final Completer<ImageCacheStatus?> completer =
        Completer<ImageCacheStatus?>();
    _createErrorHandlerAndKey(
      configuration,
      (T key, ImageErrorListener innerHandleError) {
        completer
            .complete(PaintingBinding.instance!.imageCache!.statusForKey(key));
      },
      (T? key, Object exception, StackTrace? stack) async {
        if (handleError != null) {
          handleError(exception, stack);
        } else {
          InformationCollector? collector;
          assert(() {
            collector = () => <DiagnosticsNode>[
                  DiagnosticsProperty<ImageProvider>('Image provider', this),
                  DiagnosticsProperty<ImageConfiguration>(
                      'Image configuration', configuration),
                  DiagnosticsProperty<T>('Image key', key, defaultValue: null),
                ];
            return true;
          }());
          FlutterError.reportError(FlutterErrorDetails(
            context: ErrorDescription(
                'while checking the cache location of an image'),
            informationCollector: collector,
            exception: exception,
            stack: stack,
          ));
          completer.complete(null);
        }
      },
    );
    return completer.future;
  }

  /// This method is used by both [resolve] and [obtainCacheStatus] to ensure
  /// that errors thrown during key creation are handled whether synchronous or
  /// asynchronous.
  void _createErrorHandlerAndKey(
    ImageConfiguration configuration,
    _KeyAndErrorHandlerCallback<T> successCallback,
    _AsyncKeyErrorHandler<T?> errorCallback,
  ) {
    T? obtainedKey;
    bool didError = false;
    Future<void> handleError(Object exception, StackTrace? stack) async {
      if (didError) {
        return;
      }
      if (!didError) {
        errorCallback(obtainedKey, exception, stack);
      }
      didError = true;
    }

    // If an error is added to a synchronous completer before a listener has been
    // added, it can throw an error both into the zone and up the stack. Thus, it
    // looks like the error has been caught, but it is in fact also bubbling to the
    // zone. Since we cannot prevent all usage of Completer.sync here, or rather
    // that changing them would be too breaking, we instead hook into the same
    // zone mechanism to intercept the uncaught error and deliver it to the
    // image stream's error handler. Note that these errors may be duplicated,
    // hence the need for the `didError` flag.
    final Zone dangerZone = Zone.current.fork(
      specification: ZoneSpecification(
        handleUncaughtError: (Zone zone, ZoneDelegate delegate, Zone parent,
            Object error, StackTrace stackTrace) {
          handleError(error, stackTrace);
        },
      ),
    );
    dangerZone.runGuarded(() {
      Future<T> key;
      try {
        key = obtainKey(configuration);
      } catch (error, stackTrace) {
        handleError(error, stackTrace);
        return;
      }
      key.then<void>((T key) {
        obtainedKey = key;
        try {
          successCallback(key, handleError);
        } catch (error, stackTrace) {
          handleError(error, stackTrace);
        }
      }).catchError(handleError);
    });
  }

  /// Called by [resolve] with the key returned by [obtainKey].
  ///
  /// Subclasses should override this method rather than calling [obtainKey] if
  /// they need to use a key directly. The [resolve] method installs appropriate
  /// error handling guards so that errors will bubble up to the right places in
  /// the framework, and passes those guards along to this method via the
  /// [handleError] parameter.
  ///
  /// It is safe for the implementation of this method to call [handleError]
  /// multiple times if multiple errors occur, or if an error is thrown both
  /// synchronously into the current part of the stack and thrown into the
  /// enclosing [Zone].
  ///
  /// The default implementation uses the key to interact with the [ImageCache],
  /// calling [ImageCache.putIfAbsent] and notifying listeners of the [stream].
  /// Implementers that do not call super are expected to correctly use the
  /// [ImageCache].
  @protected
  void resolveStreamForKey(ImageConfiguration configuration, ImageStream stream,
      T key, ImageErrorListener handleError) {
    // This is an unusual edge case where someone has told us that they found
    // the image we want before getting to this method. We should avoid calling
    // load again, but still update the image cache with LRU information.
    if (stream.completer != null) {
      final ImageStreamCompleter? completer =
          PaintingBinding.instance!.imageCache!.putIfAbsent(
        key,
        () => stream.completer!,
        onError: handleError,
      );
      assert(identical(completer, stream.completer));
      return;
    }
    final ImageStreamCompleter? completer =
        PaintingBinding.instance!.imageCache!.putIfAbsent(
      key,
      () => load(key, PaintingBinding.instance!.instantiateImageCodec),
      onError: handleError,
    );
    if (completer != null) {
      stream.setCompleter(completer);
    }
  }

  /// Evicts an entry from the image cache.
  ///
  /// Returns a [Future] which indicates whether the value was successfully
  /// removed.
  ///
  /// The [ImageProvider] used does not need to be the same instance that was
  /// passed to an [Image] widget, but it does need to create a key which is
  /// equal to one.
  ///
  /// The [cache] is optional and defaults to the global image cache.
  ///
  /// The [configuration] is optional and defaults to
  /// [ImageConfiguration.empty].
  ///
  /// {@tool snippet}
  ///
  /// The following sample code shows how an image loaded using the [Image]
  /// widget can be evicted using a [NetworkImage] with a matching URL.
  ///
  /// ```dart
  /// class MyWidget extends StatelessWidget {
  ///   const MyWidget({
  ///     Key? key,
  ///     this.url = ' ... ',
  ///   }) : super(key: key);
  ///
  ///   final String url;
  ///
  ///   @override
  ///   Widget build(BuildContext context) {
  ///     return Image.network(url);
  ///   }
  ///
  ///   void evictImage() {
  ///     final NetworkImage provider = NetworkImage(url);
  ///     provider.evict().then<void>((bool success) {
  ///       if (success) {
  ///         debugPrint('removed image!');
  ///       }
  ///     });
  ///   }
  /// }
  /// ```
  /// {@end-tool}
  Future<bool> evict(
      {ImageCache? cache,
      ImageConfiguration configuration = ImageConfiguration.empty}) async {
    cache ??= imageCache;
    final T key = await obtainKey(configuration);
    return cache!.evict(key);
  }

  /// Converts an ImageProvider's settings plus an ImageConfiguration to a key
  /// that describes the precise image to load.
  ///
  /// The type of the key is determined by the subclass. It is a value that
  /// unambiguously identifies the image (_including its scale_) that the [load]
  /// method will fetch. Different [ImageProvider]s given the same constructor
  /// arguments and [ImageConfiguration] objects should return keys that are
  /// '==' to each other (possibly by using a class for the key that itself
  /// implements [==]).
  Future<T> obtainKey(ImageConfiguration configuration);

  /// Converts a key into an [ImageStreamCompleter], and begins fetching the
  /// image.
  ///
  /// The [decode] callback provides the logic to obtain the codec for the
  /// image.
  ///
  /// See also:
  ///
  ///  * [ResizeImage], for modifying the key to account for cache dimensions.
  @protected
  ImageStreamCompleter load(T key, DecoderCallback decode);

  @override
  String toString() => '${objectRuntimeType(this, 'ImageConfiguration')}()';
}
